Rcellforest/R/seurat_loader.R

```diff
@@ -0,0 +1,114 @@
+#' @include example_spec.R
+#' @include helper_functions.R
+
+PCA_EMBED_KEY = "pca"
+UMAP_EMBED_KEY = "umap"
+
+#' @title Load RDS matrix and embeddings at process run
+#'
+#' @description Loads RDS from most recent layer (including current layer).
+#' If RDS comes from parental folder, append embeddings from the current layer.
+#'
+#' @importFrom reticulate import
+#' @importFrom Seurat CreateDimReducObject
+#' @importFrom glue glue
+#'
+#' @param cf_branch CellForest branch at selected process
+#'
+#' @return Seurat object with cached dimensionality reduction embeddings
+#'
+#' @section Installation:
+#' 1. `setwd(PATH_TO_CELLFOREST)`
+#' 2. `install(Rcellforest)`
+#'
+#' @export
+#'
+#' @examples
+#' library(reticulate)
+#' library(Seurat)
+#'
+#' library(Rcellforest)
+#' cellforest <- import("cellforest")
+#'
+#' root_dir <- "tests/data/example_usage/root"
+#' cf_branch <- cellforest$load(root_dir, spec = example_spec_r)
+#' cf_branch$goto_process("reduce")  # put process or alias (if exists)
+#' seurat_obj <- get_seurat_object(cf_branch)
+#'
+#' DimPlot(seurat_obj, reduction = "pca")
+get_seurat_object <- function(cf_branch) {
+  current_process <- cf_branch$current_process
+  current_path_map <- cf_branch[current_process]$path_map
+  rds_path <- toString(current_path_map$rna_r)
+  seurat_object <- readRDS(file = rds_path)
+  print(toString(glue("Creating Seurat object at process '{current_process}'"))); print(date())
+
+  # check if rds is located in the same folder as metadata (if not -> needs update)
+  meta_path <- toString(current_path_map$meta)
+  rds_path_prefix <- get_prefix_from_path(rds_path)
+  meta_path_prefix <- get_prefix_from_path(meta_path)
+
+  if (rds_path_prefix != meta_path_prefix) {
+    # TO-DO: is there a prettier way to get precursors for current process?
+    precursors <- cf_branch$spec$get_precursors_lookup(incl_current = TRUE)[[current_process]]
+    for (process_name in precursors) {
+      process <- cf_branch[process_name]$process
+      if (process == "reduce") {
+        seurat_object <- add_dim_reduc_embed(
+          seurat_object,
+          current_path_map
+        )
+      }
+
+      # TO-DO: Add loading for new processes
+    }
+  }
+  print(toString(glue("Seurat object at process '{current_process}' created"))); print(date())
+
```

Comment:
Yea, for now I'm loading UMAP from the metadata file. Will add loading information into the metadata slot. Q: for PCA, do you plan to put it in meta later on? I can pre-write code so it can get PCA from meta.

Reply:
Yeah, we should add it to metadata at some point. As long as it works currently, though, no rush

Reply:
For now, I will load in metadata at each step starting from where the RDS is until the selected process. Outside of that, adding UMAP and PCA embeds. I don't expect the latter to change, only the path from which they are taken might change. Other than that, they still use `CreateDimReducObejct()` instead of `AddMetaData()` :)